### PR TITLE
RFC Indices and finer grained events

### PR DESCRIPTION
**Pull request by rshriram**
Here is my take on the data structures:
1. the file in this PR has a list of maps and back pointers to owning objects (I didn't care much about idiomatic go). I have also added conditions during which endpoints/clusters/routes/listeners will change with respect to the service registry
2. The focus is on cutting down redundant computation and trimming memory usage. So the tradeoff here has been to dynamically compose the xDS response using proto bits from the hash map, instead of caching the entire proto/json response to a single endpoint's XDS call.
3. With this fine grained information, we could have a more cleaner invalidation or update system
(i.e. in the AppendServiceHandler (which responds to changes in services), and in AppendInstanceHandler (which responds to changes in endpoints)) 
Instead of blowing up the entire cache as done here https://github.com/istio/istio/blob/master/pilot/pkg/proxy/envoy/v1/discovery.go#L330-L337
(`out.clearcache()`), we could use the information to selectively update the right parts (like listeners/clusters/etc/). Essentially, there is no cache. The cache is the hashmaps that we have. And the cache invalidation is updating parts of the hash map. We get to reuse the code from serviceregistry/model, etc.

Thoughts? (@ZackButcher / @costinm / @kyessenov / @ijsnellf / @GregHanson / @rosenhouse ) ?

### CODE REVIEW

File: pilot/pkg/proxy/envoy/v2/datastructures.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	virtualHosts []*BoxedVirtualHost
+	// list of listeners that use this route configuration
+	listeners []*xdsapi.Listener
+}
+
+// BoxedEndpoint holds the endpoint proto representation and pointers to
+// owning clusters
+type BoxedEndpoint struct {
+	// The proto representation of this endpoint
+	protoEndpoint *xdsapi.LbEndpoint
```
Comment (kyessenov) [inline pilot/pkg/proxy/envoy/v2/datastructures.go]: Endpoint already has labels, maybe use it instead?

File: pilot/pkg/proxy/envoy/v2/datastructures.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	// list of clusters that contain this endpoint
+	clusters []*BoxedCluster
+	// list of labels associated with this endpoint
+	labels []*model.Labels
+}
+
+// BoxedCluster holds the cluster proto representation, pointers to endpoints
+type BoxedCluster struct {
```
Comment (kyessenov) [inline pilot/pkg/proxy/envoy/v2/datastructures.go]: Merged the two into ClusterLoadAssignment?

File: pilot/pkg/proxy/envoy/v2/datastructures.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	endpoints []*BoxedEndpoint
+	// List of labels associated with this cluster
+	labels []*model.Labels
+}
+
+// Indices is a set of reverse indices to quickly locate endpoints, clusters, routes
+// and listeners
+type Indices struct {
+	// Mutex guards guarantees consistency of updates to members shared across threads.
+	mu sync.RWMutex
```
Comment (kyessenov) [inline pilot/pkg/proxy/envoy/v2/datastructures.go]: global mutex per all-nodes seems to be a bottleneck.
How do we plan to stage config rollout here? Shouldn't we keep both old and new configs for some time?

File: pilot/pkg/proxy/envoy/v2/datastructures.go (hunk 2 above)
Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/datastructures.go]: If all fields are used in both cases - maybe. Otherwise - 8 bytes per pointer...

File: pilot/pkg/proxy/envoy/v2/datastructures.go (hunk 3 above)
Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/datastructures.go]: It's not global if per node. 

Not sure what is best - and don't remember memory cost per mutex, but usually easier to move to global lock than reverse.

**Comment by costinm** [comment]
One factor to consider is that in v2 we'll also have a map of active envoy configs, with the pending response. Since it's long poll, we can update only the structs we need, and it acts as a cache ( in particular for streaming grpc)

**Comment by kyessenov** [comment]
@costinm is there a design doc for that? What does "v2" refer here and what is this long poll?

File: pilot/pkg/proxy/envoy/v2/datastructures.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+	xdsapi "github.com/envoyproxy/go-control-plane/api"
+	"istio.io/istio/pilot/pkg/model"
+)
+
+// Endpoints (react to updates in registry endpoints) (notify clusters)
+// 1. Endpoints can be shared across multiple clusters.
+// 2. The membership of an endpoint in a cluster is based on the set of labels
+//    associated with the endpoint and the set of label selectors associated with the cluster.
+// 3. When a new endpoint is added, the label selectors associated with the clusters have
```
Comment (costinm) [inline pilot/pkg/proxy/envoy/v2/datastructures.go]: How ? Linear search in the clusters first, and later more advanced if needed ?

Also: once it's added, do we push anything to the cluster, or wait for next EDS call ?

File: pilot/pkg/proxy/envoy/v2/datastructures.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+//    the same listener port.
+// 5. Listeners are removed when there are no services/gateways/extsvcs listening on the listener
+//    port.
+
+// The indices here are doubly linked. A pair of objects are dependent on each other if they have
+// to be updated when one changes.
+
+// BoxedVirtualHost holds the virtual host proto representation and pointers to
+// owning route configurations.
+type BoxedVirtualHost struct {
```
Comment (costinm) [inline pilot/pkg/proxy/envoy/v2/datastructures.go]: 'boxed' is another term that may be confusing - but may be fine.

File: pilot/pkg/proxy/envoy/v2/datastructures.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+	listeners []*xdsapi.Listener
+}
+
+// BoxedEndpoint holds the endpoint proto representation and pointers to
+// owning clusters
+type BoxedEndpoint struct {
+	// The proto representation of this endpoint
+	protoEndpoint *xdsapi.LbEndpoint
+	// list of clusters that contain this endpoint
+	clusters []*BoxedCluster
```
Comment (costinm) [inline pilot/pkg/proxy/envoy/v2/datastructures.go]: General comment - can you explain when slices are used and when maps are used ?
Do we expect to never need to do O(1) lookup and only iterate ?

File: pilot/pkg/proxy/envoy/v2/datastructures.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+// BoxedCluster holds the cluster proto representation, pointers to endpoints
+type BoxedCluster struct {
+	// The proto representation of this cluster
+	protoCluster *xdsapi.Cluster
+	// list of endpoints associated with this cluster
+	endpoints []*BoxedEndpoint
+	// List of labels associated with this cluster
+	labels []*model.Labels
```
Comment (costinm) [inline pilot/pkg/proxy/envoy/v2/datastructures.go]: Labels in particular may be better as a map. Don't remember what's inside model.Label, but pointers to structs are sometimes more expensive than a bit of duplicated local data.

File: pilot/pkg/proxy/envoy/v2/datastructures.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+	// Map of virtual hosts indexed by name (not DNS). When a route rule has source based/gateway
+	// based match, the virtual host name has to be tweaked accordingly.
+	virtualHosts map[string]*BoxedVirtualHost
+
+	// Map of route configurations (used by RDS in each listener) indexed by name.
+	routeConfigs map[string]*BoxedRouteConfiguration
+
+	// Map of clusters indexed by name
+	clusters map[string]*BoxedCluster
```
Comment (costinm) [inline pilot/pkg/proxy/envoy/v2/datastructures.go]: clusterByName ?

**Comment by costinm** [comment]
Design doc for what ? Long poll is what Envoy is doing in v2/REST
(apparently for both gRPC fetch and json), pilot keeps request waiting
until there is
an update.

For the data structures - I think the comments are in effect the design, no
need to copy them to a doc that will be out of date.

All 'long poll' implementations I know in go have a map of 'active'
requests and push to them.

**Comment by costinm** [comment]
https://en.wikipedia.org/wiki/Push_technology - AFAIK envoy json is pretty
good match for what is typically called long polling.

XDS_PROTOCOL.md uses the same term:
"Synchronous (long) polling via REST endpoints is also available for the
xDS singleton APIs"

File: pilot/pkg/proxy/envoy/v2/datastructures.go (hunk 3 above)
Comment (rshriram) [reply pilot/pkg/proxy/envoy/v2/datastructures.go]: Ignore the mutex please. It was a copy paste thing from an old struct definition. I wouldn’t have such a coarse grained mutex for this entire struct.

**Comment by costinm** [comment]
Well, you probably need a mutex for the big struct for sure.

Too fine-grained muxtexes may seem as an optimization - but sometimes they
hurt more then help ( if you consider low-level cpu behavior).
Agree we should ignore mutexes for now - once we have implementation and
bench tests we can evaluate.

**Comment by kyessenov** [comment]
@costinm I am familiar what the long poll is, I was merely referring to your authoritative statement: `in v2 we'll also have a map of active envoy configs, with the pending response`, which seems like 1) some decision has been made to use a long poll, 2) there is some plan to do that already.

File: pilot/pkg/proxy/envoy/v2/datastructures.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+//    to be evaluated to determine the clusters to which the endpoint belongs to.
+// 4. When an endpoint is deleted, it needs to be removed from all associated clusters.
+// 5. When one or more labels are added to an endpoint, the label selectors from clusters
+//    that "do not have this endpoint" need to be reevaluated to see if the newly added labels
+//    would cause this endpoint to be added to these clusters.
+// 6. When one or more labels are "removed" from an endpoint, the clusters currently associated
+//    with the endpoint need to be reevaluated to see if the endpoint still qualifies to be a member
+//    of these clusters.
+
+// Clusters (react to updates in services, dest rules, ext services)
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply pilot/pkg/proxy/envoy/v2/datastructures.go]: Given that there is a cluster subset in Envoy, using the term "subset" for an Envoy "cluster" is also confusing.

**Comment by kyessenov** [comment]
@costinm Thanks for explaining. Yes, you are right, even ADS is essentially a long poll over gRPC bi-di in envoy. I was confused since I wasn't sure whether you are referring to changing legacy poll to long poll or to bi-di long poll. In any case, let's distract from the topic of this PR, which is only tangentially related to config transport mechanism.

**Comment by kyessenov** [comment]
@rshriram I find it easier to reason about efficiency of a data structure if we know the interface that the data structure implements. The indexes should be designed around the externally facing operations. Worst comes to worst, we could take an off-the-shelf database to optimize hashmaps, but the interface operations are a lot harder to change.

**Comment by costinm** [comment]
Long poll is closely related - it implies pilot will have a map with all
active sidecars that connect to this pilot, which can link to those structs
- and back. It also has implications on caching - and on copy. Once a route
proto is serialized, it can be sent with 0 copy to all connected envoys,
and may not need to be saved.

It is important to be aware of the larger picture...

**Comment by costinm** [comment]
Kuat - interfaces and abstractions have a role, but this PR is about
efficient data structures and minimal memory,  O(1) operations, and as
direct as possible implementation of the envoy interface.

File: pilot/pkg/proxy/envoy/v2/datastructures.go (hunk 2 above)
Comment (rshriram) [reply pilot/pkg/proxy/envoy/v2/datastructures.go]: ClusterLoadAssignment is the EDS response but not the CDS one. and Cluster object is the thing that has all the cluster config.

**Comment by rshriram** [comment]
@costinm 
> One factor to consider is that in v2 we'll also have a map of active envoy configs, with the pending response. Since it's long poll, we can update only the structs we need, and it acts as a cache ( in particular for streaming grpc)

We have this sorted out already (or atleast a PoC) in go-control-plane, where we have the concept of a snapshot (an envoy config). You can set the snapshot in one operation and then the ADS server will take care of sequencing the updates to the consumer envoy.

**Comment by rshriram** [comment]
updates:
1. Some finer grained locks
2. Added a rough sketch of how I envision this to work (by plumbing this into the model events and the ADS server). The proto struct copying is very bad, and I could use some help in fixing the go cp/finding work arounds to avoiding this copy
3. The ADS server could be very easily replaced by a REST server as well, and the code would not change much.

Essentially, we construct listeners/clusters/routes once when something changes, and ship the same stuff to *every* endpoint (we have to fine tune this logic to trim clusters and routes based on source match).

@kyessenov I believe this answers the interface questions. The stuff in go-cp could use a serious renaming pass, which I will get to soon.

File: pilot/pkg/proxy/envoy/v2/datastructures.go (hunk 5 above)
Comment (rshriram) [reply pilot/pkg/proxy/envoy/v2/datastructures.go]: I am open for a better word personally, but couldn't find one.. Essentially, this is a wrapper over the proto struct, wherein I am just adding some pointers to other related things.

File: pilot/pkg/proxy/envoy/v2/datastructures.go (hunk 1 above)
Comment (rshriram) [reply pilot/pkg/proxy/envoy/v2/datastructures.go]: It has metadata. I checked it.. its a map of key and struct!.. It will result in more verbose code

File: pilot/pkg/proxy/envoy/v2/datastructures.go [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+// 1. Every service (registry defined, external) in the system has a corresponding cluster
+// 2. Every subset expressed in a DestinationRule will have an associated cluster.
+// 3. Clusters are added/removed when a service is added/removed or when subsets are
+//    added/removed from DestinationRule or when a service is updated (ports are added/removed)
+// 4. Clusters need to be updated when new settings (e.g., circuit breakers, auth) are added
```
[the remark this answers is not in the document]
Comment (rshriram) [reply pilot/pkg/proxy/envoy/v2/datastructures.go]: Yep.

[the remark this answers is not in the document]
Comment (rshriram) [reply pilot/pkg/proxy/envoy/v2/datastructures.go]: What do you mean connected? All endpoints connected to Pilot. This update propagation has to be staggered.

File: pilot/pkg/proxy/envoy/v2/datastructures.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+//    that "do not have this endpoint" need to be reevaluated to see if the newly added labels
+//    would cause this endpoint to be added to these clusters.
+// 6. When one or more labels are "removed" from an endpoint, the clusters currently associated
+//    with the endpoint need to be reevaluated to see if the endpoint still qualifies to be a member
+//    of these clusters.
+
+// Clusters (react to updates in services, dest rules, ext services)
+// 1. Every service (registry defined, external) in the system has a corresponding cluster
+// 2. Every subset expressed in a DestinationRule will have an associated cluster.
+// 3. Clusters are added/removed when a service is added/removed or when subsets are
```
[the remark this answers is not in the document]
Comment (rshriram) [reply pilot/pkg/proxy/envoy/v2/datastructures.go]: I already noted the subsets thing.. Yes, if a new service is added on a port that does not exist already, then new listeners are needed.. This LDS push goes to *all* nodes, because in kube, its all to all networking.

File: pilot/pkg/proxy/envoy/v2/datastructures.go (hunk 6 above)
Comment (rshriram) [reply pilot/pkg/proxy/envoy/v2/datastructures.go]: good point.. added copious comments.

File: pilot/pkg/proxy/envoy/v2/datastructures.go (hunk 7 above)
Comment (rshriram) [reply pilot/pkg/proxy/envoy/v2/datastructures.go]: agreed.

File: pilot/pkg/proxy/envoy/v2/datastructures.go (hunk 8 above)
Comment (rshriram) [reply pilot/pkg/proxy/envoy/v2/datastructures.go]: envoyClusters?

**Comment by rshriram** [comment]
With regard to the existing mesh.go - I find it to be a mix of functionalities for multi-zone mesh and pilot optimizations. I would like to focus solely on pilot optimizations and keep the multi-zone mesh separate. Mixing both leads to complex code.

It is also duplicating quite a bit of functionality that already exists (e.g., config change events, new platform registry interfaces, etc.). 
- The existing config watcher system (in model) and the set of event handlers has been tried and tested for a year, while the one in mesh.go is similar yet limited (it is not handling route rules, external services, gateways, etc.). I am unclear on the need to reinvent that piece.
- I am unsure of the need to rewrite all of pilot registry adapters. Things like this:

```go
// NewEndpoint is a boiler-plate function intended for environment specific registries to create a new Endpoint.
func NewEndpoint(address string, port uint32, socketProtocol SocketProtocol, labels []EndpointLabel) (*Endpoint, error) {
```

essentially ignores all the platform adapters that we have been using/testing for a year. It will require all platform adapters to be rebuilt causing more code churn than necessary, and in the end we will be deploying even more code that is less tested. 

The memory/perf gains are from the data structures and the code here is a super set of the data structures in mesh.go, with lesser number of hash map lookups (aka more tlb friendly). 

This code also uses the ADS server that we (istio team) asked the rest of the community to use (in go-control-plane). There are some gaps there (such as those unnecessary dispatches that you (costin) pointed out), which need to be fixed, but I am sure @kyessenov would be happy to squeeze every last ounce of performance from that library, which would benefit other users of that library as well.

I think if we add a REST server to that library, our code would be simpler.

**Comment by sdake** [comment]
@rshriram would you mind responding on this thread re next steps for multicloud: https://groups.google.com/forum/#!topic/istio-environments/3SZaZGhPk0U

I want to get moving on multicloud.  The current churn is typical, but would be helpful to have a path forward.

Cheers
-steve

**Comment by rshriram** [comment]
I don't think this is solely about performance. This is about the system design and how to structure the code, with comments that clearly explain the control/data flow, and *not reinventing the wheel*. It actually shows how to leverage most parts of pilot in the registry while reworking the underlying data structures.

**Comment by gnirodi** [comment]
So for the benefit of @sdake and myself, could you please let us know how this integrates with v2/mesh.go?

**Comment by rshriram** [comment]
On Mon, Feb 12, 2018 at 5:02 PM Gautam Nirodi <notifications@github.com>
wrote:

> So for the benefit of @sdake <https://github.com/sdake> and myself, could
> you please let us know how this integrates with v2/mesh.go?
>
>
Mesh.go (multi cloud) is rapidly evolving and will undergo simplifications
over time.  However, I am unwilling to let that complexity and evolution
permeate all of Pilot, as it seems to be doing today.

This code is meant to cleanly separate what is a v2 implementation from
what is a multicloud thing. I am trying to keep large parts of pilot code
intact in an attempt to reach 1.0. Reimplementing the platform registry
interfaces, config event system etc as done by mesh.go seems
counterintuitive to reaching that goal.

The crux of the problem is the act of clearing the caches when platform
registry changes. That can be handled more easily as I have shown here. The
label handling can be optimized separately as well.

**Comment by rshriram** [comment]
closing as it does not align with work on multicloud.

**Comment by baodongli** [comment]
@rshriram: I am trying to get a handle on this proposal. I can see that it makes a lot of sense.
 
 A few questions: code in proxy/envoy/v1 will not be much of use for v2 with roughly the same design as in this PR, is that a fair assessment? I can imagine some changes in other parts of pilot to support this design as well, do you agree?

If I understand correctly, We'll end up having these caches in pilot:
  C0: caches possibly maintained by the underlying api server client (such as k8s go-client)
  C1: caches maintained in individual service registries as what exists today. 
  C2: caches as proposed in this PR: endpoints, clusters, routes, listeners that are lined up with envoy V2 quite well. 
  C3: snapshots (Corresponds to the existing cached responses in v1) tailored to envoy v2 

I can imagine with this proposal, it can cut down the duplicate processing in building listeners, clusters, routes, etc that exist in proxy/envoy/v1 code today. By duplicate processing, I mean the logic that will be run for all the first requests (or when changes take place in platform registries as you pointed out in your above comments) although it'd result in the same results. And if done right, it can make it quite efficient. Some optimization in C1 might be useful. E.g., the big loop in GetSidecarServiceInstances(). But it's implementation details in terms of how C2 interacts with C1 for this design, or the methods required of C1 for V2 might be different from what's there. And it  might even be possible eventually C1 could be optimized out for V2. 

To complete this design, details might need to be fleshed out as to how to build responses for xDS requests that come from individual envoy proxies. I can see that the majority parts of the responses have already been built in cache C2. What remains to be built is the parts that are specific to the sidecar that the requests are coming from, which is then combined with what's in the Cache C2, to have the complete response built. Let me know if this understanding is correct.

.

**Comment by rshriram** [comment]
FYI I have decided to close this PR as it does not align with the work that @gnirodi is doing on the multi-cloud (mesh.go).

Answers here are just for posterity.

> A few questions: code in proxy/envoy/v1 will not be much of use for v2 with roughly the same design as in this PR, is that a fair assessment? 

Not exactly. The idea is to reuse the control flow in proxy/envoy/v1 (all the function call chain as is, and simply replace filling in the json structs with filling in proto structs).

> I can imagine some changes in other parts of pilot to support this design as well, do you agree?

The idea behind this design is to keep code in pilot/pkg/serviceregistry/ stable, as well as code in pilot/pkg/model/ . The only code that will change is the code in pilot/pkg/envoy/v2/, and tests. The golden files will still be there (with different json schemas (v2 schema based on proto) )

>If I understand correctly, We'll end up having these caches in pilot:
>C0: caches possibly maintained by the underlying api server client (such as k8s go-client)

This is k8s client specific. Consul/eureka does not have C0

> C1: caches maintained in individual service registries as what exists today.

We don't have caches in the service registries. The interfaces defined in model will remain as it is.

> C2: caches as proposed in this PR: endpoints, clusters, routes, listeners that are lined up with envoy V2 quite well.

This is actually the only cache we have. In earlier model, we had per proxy response cache (that cached the entire JSON). This model has a shared cache that is finer grained.

> C3: snapshots (Corresponds to the existing cached responses in v1) tailored to envoy v2

 Yes. And this is actually just an array of pointers to the protos that are stored in C2. So its not much of a cache per se.

> I can imagine with this proposal, it can cut down the duplicate processing in building listeners, clusters, routes, etc that exist in proxy/envoy/v1 code today. By duplicate processing, I mean the logic that will be run for all the first requests (or when changes take place in platform registries as you pointed out in your above comments) although it'd result in the same results. 

Good observation. I am very happy to see that you caught on to this concept!

> And if done right, it can make it quite efficient. Some optimization in C1 might be useful. E.g., the big loop in GetSidecarServiceInstances(). But it's implementation details in terms of how C2 interacts  with C1 for this design, or the methods required of C1 for V2 might be different from what's there. And it might even be possible eventually C1 could be optimized out for V2.

As I stated above, my intent is not to rewrite all of pilot, if we are shooting for a stable 1.0. The interfaces in C1 (platform registries) will stay as it is today. It is very hard to populate three level deep envoy proto struct with information from the platform registries. We could revisit this at some point, once we get the v2 xDS servers stable.

> To complete this design, details might need to be fleshed out as to how to build responses for xDS requests that come from individual envoy proxies. I can see that the majority parts of the responses have already been built in cache C2. What remains to be built is the parts that are specific to the sidecar that the requests are coming from, which is then combined with what's in the Cache C2, to have the complete response built. Let me know if this understanding is correct.

Yep. Spot on. Thats the only part that needs to be worked out. Once we build the responses, the snapshots will contain a list of pointers to the right proto structs for a given envoy.

**Comment by costinm** [comment]
+istio-environments

On Tue, Feb 13, 2018 at 8:29 AM, Shriram Rajagopalan <
notifications@github.com> wrote:

> FYI I have decided to close this PR as it does not align with the work
> that @gnirodi <https://github.com/gnirodi> is doing on the multi-cloud
> (mesh.go).
>
I still believe it is important for @gnirodi and other people working on
multicloud to check this PR and incorporate
parts of it into the new data structures, and I would expect optimizations
on the routing side (v1alpha2) will
use something similar.



> Answers here are just for posterity.
>
> A few questions: code in proxy/envoy/v1 will not be much of use for v2
> with roughly the same design as in this PR, is that a fair assessment?
>
> Not exactly. The idea is to reuse the control flow in proxy/envoy/v1 (all
> the function call chain as is, and simply replace filling in the json
> structs with filling in proto structs).
>
The short term plan - yes, of course, we need to iterate and get something
working so it can be fine tuned.

Longer term, and based on scale test results - we may replace parts of the
chain related to endpoints.
And as LDS/RDS/CDS get moved to v2 - I expect similar optimizations and
cleanups to be done, but
also gradually and safely.


> I can imagine some changes in other parts of pilot to support this design
> as well, do you agree?
>
> The idea behind this design is to keep code in pilot/pkg/serviceregistry/
> stable, as well as code in pilot/pkg/model/ . The only code that will
> change is the code in pilot/pkg/envoy/v2/, and tests. The golden files will
> still be there (with different json schemas (v2 schema based on proto) )
>
> If I understand correctly, We'll end up having these caches in pilot:
> C0: caches possibly maintained by the underlying api server client (such
> as k8s go-client)
>
> This is k8s client specific. Consul/eureka does not have C0
>

One of the proposals I had was to explore the lighter k8s client, which is
also no using C0
and appears better suited for us. A good (and safe) start would be to
explore its use for
fetching workloads/endpoints - as an alternative path to fill in the
multi-cloud structures.



> C1: caches maintained in individual service registries as what exists
> today.
>
> We don't have caches in the service registries. The interfaces defined in
> model will remain as it is.
>


> C2: caches as proposed in this PR: endpoints, clusters, routes, listeners
> that are lined up with envoy V2 quite well.
>
> This is actually the only cache we have. In earlier model, we had per
> proxy response cache (that cached the entire JSON). This model has a shared
> cache that is finer grained.
>
Well - the v1 implementation has C3, in the new implementation we'll only
have C2.



> C3: snapshots (Corresponds to the existing cached responses in v1)
> tailored to envoy v2
>
> Yes. And this is actually just an array of pointers to the protos that are
> stored in C2. So its not much of a cache per se.
>
> I can imagine with this proposal, it can cut down the duplicate processing
> in building listeners, clusters, routes, etc that exist in proxy/envoy/v1
> code today. By duplicate processing, I mean the logic that will be run for
> all the first requests (or when changes take place in platform registries
> as you pointed out in your above comments) although it'd result in the same
> results.
>
> Good observation. I am very happy to see that you caught on to this
> concept!
>

I think it should go a bit further: since v2 is a push API, this logic
should also be able to process events ( endpoint
added/changed ) and directly send it to sidecars, without waiting for a
poll.

The pilot will have a map of long-polling sidecars, and one of the reasons
for switch RDS/LDS/CDS to v2 is to
propagate changes immediately (or gradually under server control), instead
of each sidecar polling every
few seconds. That's not in scope for multi-cloud, but will need to be done
in istio.



> And if done right, it can make it quite efficient. Some optimization in C1
> might be useful. E.g., the big loop in GetSidecarServiceInstances(). But
> it's implementation details in terms of how C2 interacts with C1 for this
> design, or the methods required of C1 for V2 might be different from what's
> there. And it might even be possible eventually C1 could be optimized out
> for V2.
>
> As I stated above, my intent is not to rewrite all of pilot, if we are
> shooting for a stable 1.0. The interfaces in C1 (platform registries) will
> stay as it is today. It is very hard to populate three level deep envoy
> proto struct with information from the platform registries. We could
> revisit this at some point, once we get the v2 xDS servers stable.
>
+1 - we want to iterate and keep things stable.

It is possible to have an alternate 'direct path' for endpoints, for
example using the lighter k8s client to
directly fill the endpoint structs - but with a flag and as an option, with
the default as the stable (current)
version until the new code has similar stability.

To complete this design, details might need to be fleshed out as to how to
> build responses for xDS requests that come from individual envoy proxies. I
> can see that the majority parts of the responses have already been built in
> cache C2. What remains to be built is the parts that are specific to the
> sidecar that the requests are coming from, which is then combined with
> what's in the Cache C2, to have the complete response built. Let me know if
> this understanding is correct.
>
> Yep. Spot on. Thats the only part that needs to be worked out. Once we
> build the responses, the snapshots will contain a list of pointers to the
> right proto structs for a given envoy.
>
>
>

**Comment by costinm** [comment]
https://kubernetes.io/docs/reference/client-libraries/ ->
github.com/ericchiang/k8s

I did a quick test - it is just a generated wrapper on top of the
libraries, with minimal 'opinions'.

Costin

**Comment by baodongli** [comment]
From looking at its GitHub page, it's not clear to me whether or not it supports changes of services and endpoints in the cluster.

**Comment by baodongli** [comment]
it does support watch.

**Comment by ayj** [comment]
> It is possible to have an alternate 'direct path' for endpoints, for example using the lighter k8s client to directly fill the endpoint structs - but with a flag and as an option, with the default as the stable (current) version until the new code has similar stability.

What problem is this trying to solve?

FWIW, the recommendation from https://github.com/ericchiang/k8s#should-i-use-this-or-client-go says:

> TL;DR - Use client-go if you're writing a controller.

**Comment by costinm** [comment]
On Tue, Feb 13, 2018 at 1:02 PM, Jason Young <notifications@github.com>
wrote:

> It is possible to have an alternate 'direct path' for endpoints, for
> example using the lighter k8s client to directly fill the endpoint structs
> - but with a flag and as an option, with the default as the stable
> (current) version until the new code has similar stability.
>
> What problem is this trying to solve?
>
Lower memory overhead. In multi-cluster the goal is to scale to 10s of
clusters, each with max number of endpoints.


> FWIW, the recommendation from https://github.com/ericchiang/
> k8s#should-i-use-this-or-client-go says:
>
> TL;DR - Use client-go if you're writing a controller.
>
>
We are not writing a controller.

Istio does have an ingress controller - and it should keep using client-go
- but for most of the code fetching data ( endpoints in particular )
the client-go caching and allocations may be a problem. Right now we are
seeing a pretty large GC overhead with pilot, that's one
of the things to explore - of course we're at early stage, just trying to
get heap dumps and find simpler solutions.

Costin

**Comment by costinm** [comment]
And to clarify: I am not proposing we switch the client.

Just to evaluate if it's even an option, and wait until we have more real
data on where the memory is spent.

There is another use case for the library - on agent side - since it has
fewer dependencies and lower footprint,
but I hope we can continue to avoid a dependency ( loading SNI certs
dynamically was one case for agent using
k8s API)

Costin
